### PR TITLE
Change yacc and protos for multidim key ranges.

### PR DESCRIPTION
**Pull request by reshke**

### CODE REVIEW

File: pkg/models/kr/keyrange.go [hunk 1]
```diff
@@ -1,106 +1,294 @@
...
 
-	return len(kr) < len(other)
+	return len(bound) < len(key)
 }
 
-// TODO : unit tests
-func CmpRangesLessEqual(kr []byte, other []byte) bool {
-	if len(kr) == len(other) {
-		return string(kr) <= string(other)
+func (kr *KeyRange) InFunc(attribind int, raw []byte) {
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:38]:
```suggestion
func (kr *KeyRange) InFunc(attrBind int, raw []byte) {
```

File: pkg/models/kr/keyrange.go [hunk 2]
```diff
@@ -1,106 +1,294 @@
...
+			i2 := key[i].(string)
+			if i1 == i2 {
+				// continue
+			} else if CmpRangesLessStringsDeprecated(i1, i2) {
+				return true
+			} else {
+				return false
+			}
+		default:
+			// wtf?
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:130]: we can panic here, i suppose

File: pkg/models/kr/keyrange.go [hunk 3]
```diff
@@ -1,106 +1,294 @@
...
+	return res
 }
 
 // TODO : unit tests
-func CmpRangesEqual(kr []byte, other []byte) bool {
-	if len(kr) == len(other) {
-		return string(kr) == string(other)
+func CmpRangesLess(bound KeyRangeBound, key KeyRangeBound, types []string) bool {
+	for i := 0; i < len(bound); i++ {
+		switch types[i] {
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:98]: Mb check for type match?

File: pkg/models/kr/keyrange.go [hunk 4]
```diff
@@ -1,106 +1,294 @@
...
+	}
+
+	return kr
 }
 
 // TODO : unit tests
 func (kr *KeyRange) ToDB() *qdb.KeyRange {
-	return &qdb.KeyRange{
-		LowerBound:     kr.LowerBound,
+	krqb := &qdb.KeyRange{
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:264]:
```suggestion
	krDb := &qdb.KeyRange{
```

File: pkg/models/kr/keyrange.go [hunk 5]
```diff
@@ -1,106 +1,294 @@
...
+	}
+	return krqb
 }
 
 // TODO : unit tests
 func (kr *KeyRange) ToProto() *proto.KeyRangeInfo {
-	return &proto.KeyRangeInfo{
-		KeyRange: &proto.KeyRange{
-			LowerBound: string(kr.LowerBound),
+	krprot := &proto.KeyRangeInfo{
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:280]:
```suggestion
	krProto := &proto.KeyRangeInfo{
```

**Review by EinKrebs — CHANGES_REQUESTED** [review]

**Review by EinKrebs — APPROVED** [review]

File: pkg/coord/local/clocal.go [hunk 6]
```diff
@@ -666,8 +704,13 @@ func (qr *LocalCoordinator) ListKeyRanges(ctx context.Context, distribution stri
 		return nil, err
 	} else {
 		for _, keyRange := range krs {
-			ret = append(ret, kr.KeyRangeFromDB(keyRange))
+			ds, err := qr.qdb.GetDistribution(ctx, keyRange.DistributionId)
```
Comment (EinKrebs) [inline pkg/coord/local/clocal.go:707]: You can get distribution only once per function call

File: pkg/coord/local/clocal_test.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package local_test
```
Comment (EinKrebs) [inline pkg/coord/local/clocal_test.go:1]: Unnecessary file

File: pkg/models/hashfunction/hashfunction.go [hunk 8]
```diff
@@ -31,16 +32,52 @@ var (
 // Returns:
 //   - []byte: The hashed byte slice.
```
Comment (EinKrebs) [inline pkg/models/hashfunction/hashfunction.go:33]: Change the docstring as well

File: pkg/models/kr/keyrange.go [hunk 9]
```diff
@@ -10,18 +11,22 @@ import (
...
 type KeyRange struct {
 	LowerBound   KeyRangeBound
 	ShardID      string
 	ID           string
 	Distribution string
+
+	ColumnTypes []string
 }
 
 // CmpRangesLess compares two byte slices, kr and other, and returns true if kr is less than other.
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:32]: Move this docstring to the actual CmpRangesLess method

File: pkg/models/kr/keyrange.go [hunk 10]
```diff
@@ -35,71 +40,189 @@ type KeyRange struct {
...
-//   - other: The second byte slice to compare.
-//
-// Returns:
-//   - bool: True if kr is less than or equal to other, false otherwise.
-//
-// TODO : unit tests
-func CmpRangesLessEqual(kr []byte, other []byte) bool {
-	if len(kr) == len(other) {
-		return string(kr) <= string(other)
+func (kr *KeyRange) InFunc(attribInd int, raw []byte) {
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:51]: Consider adding docstring to new methods

Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:51]: Also, maybe it's better to change functions' names to getBound/setBound

File: pkg/models/kr/keyrange.go [hunk 11]
```diff
@@ -35,71 +40,189 @@ type KeyRange struct {
...
+			i2 := key[i].(string)
+			if i1 == i2 {
+				// continue
+			} else if CmpRangesLessStringsDeprecated(i1, i2) {
+				return true
+			} else {
+				return false
+			}
+		default:
+			// wtf?
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:158]:
```suggestion
		default:
```
Comment can be dropped

File: pkg/models/kr/keyrange.go [hunk 12]
```diff
@@ -35,71 +40,189 @@ type KeyRange struct {
...
-// Returns true if kr is less than or equal to other, false otherwise.
-//
-// Parameters:
-//   - kr: The first byte slice to compare.
-//   - other: The second byte slice to compare.
-//
-// Returns:
-//   - bool: True if kr is less than or equal to other, false otherwise.
-//
-// TODO : unit tests
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:57]: Move this docstring to respective method & update it instead of deleting

File: pkg/models/kr/keyrange.go [hunk 13]
```diff
@@ -35,71 +40,189 @@ type KeyRange struct {
...
-// KeyRangeFromDB converts a qdb.KeyRange object to a KeyRange object.
-// It creates a new KeyRange object with the values from the qdb.KeyRange object.
-// It returns a pointer to the new KeyRange object.
-//
-// Parameters:
-//   - kr: The qdb.KeyRange object to convert.
-//
-// Returns:
-//   - *KeyRange: A pointer to the new KeyRange object.
-//
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:94]: Move to KeyRangeFromDB method & update

File: pkg/models/kr/keyrange.go [hunk 14]
```diff
@@ -35,71 +40,189 @@ type KeyRange struct {
...
+		case qdb.ColumnTypeVarcharDeprecated:
+			i1 := bound[i].(string)
+			i2 := key[i].(string)
+			if i1 == i2 {
+				// continue
+			} else {
+				return false
+			}
+		default:
+			// wtf?
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:196]:
```suggestion
			panic(MissTypedKeyRange)
```

File: pkg/models/kr/keyrange.go [hunk 15]
```diff
@@ -35,71 +40,189 @@ type KeyRange struct {
...
+	// TODO: Fix this! (krdb.LowerBound -> krqb.LowerBound[i])
+	// now this works only for unidim distributions
+	for i := 0; i < len(colTypes); i++ {
+		kr.InFunc(i, krdb.LowerBound[i])
 	}
+
+	return kr
 }
 
 // KeyRangeFromSQL converts a spqrparser.KeyRangeDefinition into a KeyRange.
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:228]: Update the docstring

File: pkg/models/kr/keyrange.go [hunk 16]
```diff
@@ -113,16 +236,44 @@ func KeyRangeFromDB(kr *qdb.KeyRange) *KeyRange {
...
+		ID:           krsql.KeyRangeID,
+		Distribution: krsql.Distribution,
+
+		ColumnTypes: colTypes,
+
+		LowerBound: make(KeyRangeBound, len(colTypes)),
+	}
+
+	if len(colTypes) != len(krsql.LowerBound.Pivots) {
+		return nil, fmt.Errorf("number of column mismatch with distribution")
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:254]:
```suggestion
		return nil, fmt.Errorf("number of columns mismatches with distribution")
```

File: pkg/models/kr/keyrange.go [hunk 17]
```diff
@@ -113,16 +236,44 @@ func KeyRangeFromDB(kr *qdb.KeyRange) *KeyRange {
...
+	}
+
+	for i := 0; i < len(colTypes); i++ {
+		kr.InFunc(i, val[i])
+	}
+
+	return kr
 }
 
 // KeyRangeFromProto converts a protobuf KeyRangeInfo to a KeyRange object.
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:279]: Update the docstring

File: pkg/models/kr/keyrange.go [hunk 18]
```diff
@@ -136,16 +287,24 @@ func KeyRangeFromSQL(kr *spqrparser.KeyRangeDefinition) *KeyRange {
...
+	kr := &KeyRange{
+		ShardID:      krproto.ShardId,
+		ID:           krproto.Krid,
+		Distribution: krproto.DistributionId,
+		ColumnTypes:  colTypes,
+
+		LowerBound: make(KeyRangeBound, len(colTypes)),
+	}
+
+	for i := 0; i < len(colTypes); i++ {
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:303]:
```suggestion
        if len(colTypes) != len(krsql.LowerBound.Pivots) {
		return nil, fmt.Errorf("number of columns mismatches with distribution")
        }

	for i := 0; i < len(colTypes); i++ {
```

File: pkg/models/kr/keyrange.go [hunk 19]
```diff
@@ -156,12 +315,18 @@ func KeyRangeFromProto(kr *proto.KeyRangeInfo) *KeyRange {
...
-	return &qdb.KeyRange{
-		LowerBound:     kr.LowerBound,
+	krDb := &qdb.KeyRange{
+		LowerBound:     make([][]byte, len(kr.ColumnTypes)),
 		ShardID:        kr.ShardID,
 		KeyRangeID:     kr.ID,
 		DistributionId: kr.Distribution,
 	}
+	// TODO: Fix this! (krqb.LowerBound -> krqb.LowerBound[i])
+	// now this works only for unidim distributions
```
Comment (EinKrebs) [inline pkg/models/kr/keyrange.go:325]: Unnecessary TODO

File: qdb/ops/ops.go [hunk 20]
```diff
@@ -29,7 +30,10 @@ func CreateKeyRangeWithChecks(ctx context.Context, qdb qdb.QDB, keyRange *kr.Key
 	}
 
 	for _, v := range existsKrids {
-		if kr.CmpRangesEqual(keyRange.LowerBound, v.LowerBound) {
+
+		spqrlog.Zero.Info().Bytes("types", v.LowerBound[0]).Msg("adding key range")
+		eph := kr.KeyRangeFromBytes(v.LowerBound, keyRange.ColumnTypes)
```
Comment (EinKrebs) [inline qdb/ops/ops.go:35]: Unclear variable name

**Review by EinKrebs — CHANGES_REQUESTED** [review]

File: qdb/ops/ops.go (hunk 20 above)
Comment (reshke) [reply qdb/ops/ops.go:35]: its ephemeral or temporal

File: pkg/models/kr/keyrange.go (hunk 10 above)
Comment (reshke) [reply pkg/models/kr/keyrange.go:51]: its in/out send/receive like in PG

File: pkg/models/kr/keyrange.go (hunk 13 above)
Comment (reshke) [reply pkg/models/kr/keyrange.go:94]: done

File: pkg/models/kr/keyrange.go (hunk 17 above)
Comment (reshke) [reply pkg/models/kr/keyrange.go:278]: done

File: pkg/coord/local/clocal.go (hunk 6 above)
Comment (reshke) [reply pkg/coord/local/clocal.go:706]: si

File: pkg/models/kr/keyrange.go (hunk 15 above)
Comment (reshke) [reply pkg/models/kr/keyrange.go:270]: done

File: pkg/coord/local/clocal.go [hunk 21]
```diff
@@ -477,20 +499,18 @@ func (qr *LocalCoordinator) Unite(ctx context.Context, req *kr.UniteKeyRange) er
...
-		ID:           krBase.KeyRangeID,
-	}
+	krBaseCopy := krBase
+	krBaseCopy.LowerBound = newBound
+	united := kr.KeyRangeFromDB(krBaseCopy, ds.ColTypes)
 
 	return ops.ModifyKeyRangeWithChecks(ctx, qr.qdb, united)
 }
 
+// Caller should lock key range
```
[the remark this answers is not in the document]
Comment (reshke) [reply pkg/coord/local/clocal.go:512]: sure

File: pkg/models/kr/keyrange.go (hunk 12 above)
Comment (reshke) [reply pkg/models/kr/keyrange.go:57]: done

**Review by EinKrebs — APPROVED** [review]

**Pull request merged**